### PR TITLE
Remoted reliability tests

### LINKED ISSUE #2442: Reliability test: Standarize reliability tests data sources
### Description 

Reliability tests will use as an argument a file capable of displaying in a standardized way the output of a testing environment.

In this context, it is necessary to design the structure of this file, as well as to create a script able to translate the set of artifacts (metrics, statistics, logs) into a single JSON file, which will provide relevant information to the reliability tests.

### LINKED ISSUE #2503: Remoted Reliablility Tests: Agent connection
### Description

 As part of https://github.com/wazuh/wazuh-qa/issues/2441 is necessary the creation of new reliability tests.

### Reliability tests design

There will be a suite for generic reliability tests, such as the absence of errors and warnings in logs beyond those expected.
Each module will have its own suite of tests related to specific aspects of the module.

```
tests/
└── reliability
    ├── general
    │   ├── data
    │   ├── test_general_error.py
    │   └── test_general_warnings.py
    └── remote
        ├── data
        └── test_remote_keep_alives.py
```
 
### Test design

Every test of this suite will receive two parameters:
- `config.yaml`: YAML with the environment deployment. This will be used for the test for multiple purposes:
  - Provide general data of the environment: EPS, expected generated alerts, etc.
  - Specified the object of the testing (group of agents/managers, workers, etc).
- `report.json`: Report that will describe the status of the Wazuh environment during time execution.

The expected reliability tests to develop in this iteration are:
- General
  - `test_unexpected_errors`
- Remoted
  - `test_agent_connection`
  - `test_remote_keep_alive`

### PR DESCRIPTION
**Pull request by Rebits**
|Related issue|
|---|
|#2441|

## Description

This PR includes the following developments:

- #2442 

A new script, `wazuh-report` has been created. This script will accept an artifacts folder path that will follow these structures:
```
artifacts/
├── agents
│   ├── agent1
│   │   ├── data
│   │   │   ├── binaries
│   │   │   └── stats
│   │   └── logs
│   └── agent2
│       ├── data
│       │   ├── binaries
│       │   └── stats
│       └── logs
└── managers
    ├── master
    │   └── master-instance
    │       ├── data
    │       │   ├── binaries
    │       │   └── stats
    │       └── logs
    └── workers
        └── worker1
            ├── data
            │   ├── binaries
            │   └── stats
            └── logs
```

Using these files, the `LogAnalyzer` class will generate a JSON [report](https://github.com/wazuh/wazuh-qa/issues/2442#issuecomment-1022190041), which summarizes the general status of the environment during the testing environment time.

The final user should use the `wazuh-report` script to generate this report

#### `wazuh-report` example of use

```
wazuh-report -p /tmp/testing-artifacts/ -r report.json
```
Where:

- `-p`: Artifact path
- `-r`: Report path

#### Example of report

[report.zip](https://github.com/wazuh/wazuh-qa/files/7958228/report.zip)

***

- #2503

This PR includes reliability tests for general errors and remoted/agentd daemons. The main objectives of these tests are:
- Ensure the health of the environment, no unexpected errors appear
- Agents remain connected during the environment up time:
   - Keepalives are received by the manager in the expected intervals
   - TCP sessions does not change
   - Differences between last_ack and last_keep_alive in the agents' statistics do not expect the expected maximum.

### Test design
All the tests will require a report, passed as `--report` pytest arguments.  This report should be generated with https://github.com/wazuh/wazuh-qa/issues/2442 `ReportGenerator` tool. The `test_error_messages` also, accept optional arguments:
 - `target_daemons`: Ignored error of not specified daemons
 - `target_hosts`: Ignore errors of not specified targets. This required an adapted report, that will include specific data for the target.

## Tests

- [x] Proven that tests **pass** when they have to pass.

- [x] Python codebase satisfies PEP-8 style style guide. `pycodestyle --max-line-length=120 --show-source --show-pep8 file.py`.
- [x] Python codebase is documented following the Google Style for Python docstrings.

### CODE REVIEW

File: deps/wazuh_testing/wazuh_testing/tools/sources/report_generator.py [hunk 1]
```diff
@@ -0,0 +1,371 @@
+# Copyright (C) 2015-2021, Wazuh Inc.
+# Created by Wazuh, Inc. <info@wazuh.com>.
+# This program is free software; you can redistribute it and/or modify it under the terms of GPLv2
+import os
+import re
+
+from datetime import datetime
+from itertools import groupby
+from mmap import ACCESS_READ, mmap
```
Comment (juliamagan) [inline deps/wazuh_testing/wazuh_testing/tools/sources/report_generator.py:9]:
```suggestion
import os
import re
from datetime import datetime
from itertools import groupby
from mmap import ACCESS_READ, mmap
```

File: tests/reliability/agent_connection/test_agent_connection.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+tags:
+    - remoted_agent_communication
+    - remoted
+    - agentd
+'''
+import pytest
+
+
+def test_agent_connection(get_report):
```
Comment (juliamagan) [inline tests/reliability/agent_connection/test_agent_connection.py:71]: We need to add the test documentation

File: tests/reliability/agent_connection/test_keep_alive.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    - remoted
+    - agentd
+'''
+import pytest
+
+
+MAX_DIFFERENCE_ACK_KEEP_ALIVE = 20
+
+
+def test_keep_alives(get_report):
```
Comment (juliamagan) [inline tests/reliability/agent_connection/test_keep_alive.py:73]: Same as before

File: tests/reliability/general/test_error_messages.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    pattern = re.compile(".*\d+\/\d+\/\d+ \d+:\d+:\d+ (.*?):")
+    if pattern.match(log_line):
+        return pattern.match(log_line).group(1)
+    else:
+        return None
+
+
+@pytest.mark.parametrize('code', error_codes)
+@pytest.mark.parametrize('target', target)
+def test_error_messages(get_report, code, target):
```
Comment (juliamagan) [inline tests/reliability/general/test_error_messages.py:119]: Same as before

**Review by juliamagan — CHANGES_REQUESTED** [review]
GJ, but we need to add documentation (tests, functions, fixtures...)

File: tests/reliability/agent_connection/test_agent_connection.py (hunk 2 above)
Comment (Rebits) [reply tests/reliability/agent_connection/test_agent_connection.py:71]: Done [6dc99b3](https://github.com/wazuh/wazuh-qa/pull/2612/commits/6dc99b3201c9bcca8c630ee8466bf134031fddcc)

File: tests/reliability/general/test_error_messages.py (hunk 4 above)
Comment (Rebits) [reply tests/reliability/general/test_error_messages.py:119]: Done [6dc99b3](https://github.com/wazuh/wazuh-qa/pull/2612/commits/6dc99b3201c9bcca8c630ee8466bf134031fddcc)

File: deps/wazuh_testing/wazuh_testing/tools/sources/report_generator.py [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+        """
+        timestamp_match = re.match(r"^(\d{4}/\d{2}/\d{2} \d{2}:\d{2}:\d{2}).*", line)
+        return datetime.strptime(timestamp_match.group(1), '%Y/%m/%d %H:%M:%S') if timestamp_match else None
+
+    @staticmethod
+    def findall_regex_line(lines, regex, flags=None):
+        """Get all the lines that match a certain regex.
+
+        Args:
+            Lines (list): List of lines.
```
Comment (juliamagan) [inline deps/wazuh_testing/wazuh_testing/tools/sources/report_generator.py:32]:
```suggestion
            lines (list): List of lines.
```

File: deps/wazuh_testing/wazuh_testing/tools/sources/report_generator.py [hunk 6]
```diff
@@ -0,0 +1,473 @@
...
+
+        error_lines = [error.decode(encoding='utf8') for error in error_lines] if error_lines else []
+        return error_lines
+
+    @staticmethod
+    def get_error_log_file(log_path, type='error'):
+        """Get all the lines of the specified type of the log file.
+
+        Args:
+            log_path (list): Log path-
```
Comment (juliamagan) [inline deps/wazuh_testing/wazuh_testing/tools/sources/report_generator.py:49]:
```suggestion
            log_path (list): Log path
```

File: deps/wazuh_testing/wazuh_testing/tools/sources/report_generator.py [hunk 7]
```diff
@@ -0,0 +1,473 @@
...
+            hostname (str): Hostname to get all the logs files.
+        """
+        files = os.path.join(self.get_instances_artifacts('all', hostname)[0]['path'], 'logs')
+        host_log_list = [log for log in os.listdir(files)]
+        return host_log_list
+
+    def get_instances_logs(self, log, component, hosts_regex=".*"):
+        """Get the logs path for specified hosts_regex
+
+        Args:
```
Comment (juliamagan) [inline deps/wazuh_testing/wazuh_testing/tools/sources/report_generator.py:390]: We need to add `log`

File: deps/wazuh_testing/wazuh_testing/tools/sources/report_generator.py [hunk 8]
```diff
@@ -0,0 +1,473 @@
...
+                general_path = os.path.join(host_artifact_path['path'], 'logs')
+                host_artifact_path['logs'][log_file] = os.path.join(general_path, log_file)
+            del host_artifact_path['path']
+
+        return artifacts_paths
+
+    def get_instances_process_metrics(self, process, component, hosts_regex='.*'):
+        """Get the metrics statistics path for specified hosts_regex
+
+        Args:
```
Comment (juliamagan) [inline deps/wazuh_testing/wazuh_testing/tools/sources/report_generator.py:409]: We need to add `process`

File: deps/wazuh_testing/wazuh_testing/tools/sources/report_generator.py [hunk 9]
```diff
@@ -0,0 +1,473 @@
...
+        files = self.get_instances_artifacts(component, hosts_regex)
+        for file in files:
+            file.update((host, os.path.join(artifact_path, 'data', 'binaries', process + '.csv'))
+                        for host, artifact_path in file.items())
+        return files
+
+    def get_instances_statistics(self, statistic, component, hosts_regex='.*'):
+        """Get the daemons statistics path for specified hosts_regex
+
+        Args:
```
Comment (juliamagan) [inline deps/wazuh_testing/wazuh_testing/tools/sources/report_generator.py:422]: We need to add `statistic`

File: tests/reliability/agent_connection/test_agent_connection.py [hunk 10]
```diff
@@ -0,0 +1,111 @@
+'''
+copyright: Copyright (C) 2015-2021, Wazuh Inc.
```
Comment (juliamagan) [inline tests/reliability/agent_connection/test_agent_connection.py:2]:
```suggestion
copyright: Copyright (C) 2015-2022, Wazuh Inc.
```

File: tests/reliability/agent_connection/test_keep_alive.py [hunk 11]
```diff
@@ -0,0 +1,112 @@
+'''
+copyright: Copyright (C) 2015-2021, Wazuh Inc.
```
Comment (juliamagan) [inline tests/reliability/agent_connection/test_keep_alive.py:2]:
```suggestion
copyright: Copyright (C) 2015-2022, Wazuh Inc.
```

File: tests/reliability/conftest.py [hunk 12]
```diff
@@ -0,0 +1,64 @@
+# Copyright (C) 2015-2021, Wazuh Inc.
```
Comment (juliamagan) [inline tests/reliability/conftest.py:1]:
```suggestion
# Copyright (C) 2015-2022, Wazuh Inc.
```

File: tests/reliability/general/test_error_messages.py [hunk 13]
```diff
@@ -0,0 +1,161 @@
+'''
+copyright: Copyright (C) 2015-2021, Wazuh Inc.
```
Comment (juliamagan) [inline tests/reliability/general/test_error_messages.py:2]:
```suggestion
copyright: Copyright (C) 2015-2022, Wazuh Inc.
```

**Review by juliamagan — CHANGES_REQUESTED** [review]

**Review by juliamagan — APPROVED** [review]
LGTM

**Pull request merged**